docs/guides/actors.md

```diff
@@ -332,6 +332,79 @@ parentService.send('LOCAL.WAKE');
 // => 'connected'
 ```
 
+### Stopping Spawned Machines
+
+The object returned from a successful `spawn` action is an `Interpreter`. It implements the `actor` interface, but has slightly different behavior. If you call `stop()` on an `Interpreter` it does not clean up any ongoing `activities` in the active `StateNode`. To properly stop spawned machines you must additionally send them an action using `actions.stop()`.
+
+Further, as `assign` actions are run in advance of custom user actions, by the time you would usually attempt to stop a spawned machine, you may need to take care that the reference to the machine in `context` is not overwritten.
+
+For actions which do not have side effects (such as mutating `context`), you can use `actions.pure()` to ensure that your custom actions is run in order of definition, and before `assign` actions' updates are applied to `context`. Depending upon why you need to stop a spawned machine, this technique may be required.
+
+The below machine demonstrates how to use `actions.pure()` and `actions.stop()` to properly stop a consistently respawned machine.
+
+```js {18-20,27-29,35}
+const childMachine = Machine({
+  id: 'child-machine',
+  initial: 'idle',
+  states: {
+    idle: {
+      /* ... */
+    }
+  }
+});
+
+const parentMachine = Machine({
+  id: 'stoppable-machine',
+  initial: 'spawn',
+  context: {},
+
+  states: {
+    reset: {
+      always: {
+        target: 'spawn',
+        actions: actions.pure((context) => {
+          context.childMachine.stop();
+          return actions.stop(context.childMachine.id);
```

Comment:
I included both versions of `stop` here because I *believe* that they are both required.

Reply:
I have not experienced issues, but my WIP machine is not complicated enough to trigger the thing I believe could be an issue. (And I wanted to get a first pass of the documentation together and ask before trying to create a repro of a hypothetical issue.)

My hypothesis comes out of scanning the implementations of these to get a bit more context before filing this PR. The test case that this would need is:

- Parent machine.
- Spawns a child *machine* (so that it is an `Interpreter`, not an actor).
- Child machine starts an activity (and stays in that state).
- Child machine's interpreter reference from parent machine's `context` has a listener attached to it.
- Test sending `xstate.stop`, calling `childMachine.stop()`, or doing both of those things.

Confirm:
- No activities still running. (I *think* this is what `xstate.stop` does.)
- No listeners still bound. (I *think* this is what `Interpreter#stop` does.)
- Make sure that it isn't just getting GC'd by way of clobbering all references.

I suppose that if the sent `xstate.stop` actually bubbled up into something that eventually triggered listener removal on the interpreter then it's good, but in a quick recheck I can't identify where that might happen.

Reply:
Okay, I have created the test case:

https://codesandbox.io/s/brave-fermi-nyg8e?file=/src/index.js

My findings:
- Either one of `Interpreter#stop` or `actions.stop()` seem to result in proper teardown. (One of the `stopListeners` sends a `session.stop` via `Interpreter#stop`.) 
- What I thought was happening with `activities` doesn't appear to be relevant. I'm not sure why.

I also tried to check the behavior for sending `actions.stop()` as an event to a top-level `Interpreter` which appears to not work, but I'm not entirely sure what it's doing.

@Andarist Can you also take a look at this test case yourself to make sure that my understanding is correct and that my conclusions are valid?

If my analysis holds, I believe that `Interpreter#stop`/`<actor>.stop()` is actually a complete solution.

Reply:
> @Andarist Can you also take a look at this test case yourself to make sure that my understanding is correct and that my conclusions are valid?

Sure thing! I've forked your demo so it doesn't restart the spawned child all over again - it was easier for me to inspect stuff this way: https://codesandbox.io/s/magical-lamarr-b09k1?file=/src/index.js 

> Either one of Interpreter#stop or actions.stop() seem to result in proper teardown. (One of the stopListeners sends a session.stop via Interpreter#stop.)

if we are talking about calling those from within that `pure` action then structures within child get cleaned up nicely but `context.childMachine.stop()` doesn't clean this child from the parent's structure, whereas `actions.pure((context) => actions.stop(context.childMachine.id) })` does both correctly.

> I also tried to check the behavior for sending actions.stop() as an event to a top-level Interpreter which appears to not work, but I'm not entirely sure what it's doing.

`actions.stop` doesn't create an event (well, actually it is compatible with an event interface so it could be abused, sort of, to be sent and processed). It creates an action object - think about actions as of instructions for the interpreter that you'd like something to happen. Those actions are received by the interpreter through other means though than by sending it events so to make `service.send(actions.stop())` work you would have to create a transition for the `xstate.stop` event and create a stop action there. This is not what you are looking for though, I'm just describing this for completeness. 

> // ALSO: Test the parent behavior.
> // service.stop();

This cleans up stuff correctly and I have confirmed that using this demo.

> // HELP: Is this correct for trying to send a stop action to an un-parented Interpreter?
> // service.send(actions.stop(service.id));

Do you want to stop the `service` here or some child service of a particular id? If the first one it should be totally valid to just call `stop` method like this `service.stop()`.

> // 2. Check to see if `window.childMachineFirst`'s StateNode has activities.

It seems that we do not clean children when stopping but they are being stopped. So references are held to them - gonna prepare a fix for this later as semantically it seems right the good thing to do. The child services are being stopped though so GC should kick in most of the time and collect the whole tree of services.

> // 3. Check to see if `window.childMachineFirst` has listeners.

This works correctly - those structures are being cleaned up just OK.